add apiserver-count fix proposal

rphillips opened the pull request:
This is a proposal to fix the apiserver-count issue at https://github.com/kubernetes/kubernetes/issues/22609. I would appreciate a review on the proposal.

- [x] Add ConfigMap for configurable options
- [ ] Find out dependencies on the Endpoints API and add them to the proposal

hongchaodeng commented:
Shouldn't the commit msg and title be "add apiserver-count fix proposal"? Mind the "-".

ericchiang commented:
cc @kubernetes/sig-api-machinery-proposals

luxas commented:
cc @kubernetes/sig-cluster-lifecycle-feature-requests

lpabon commented:
Hi @rphillips thanks for putting this together. Could you add the following to the documentation:

* Configurable values for timeouts or other magic numbers
* Explanation in the document on how clients consume this information

mikedanese commented on contributors/design-proposals/apiserver-count-fix.md:
```diff
@@ -0,0 +1,98 @@
+
+## Proposal
+
+### Create New Reconciler
```
This design says "create new reconciler" and has a struct definition but doesn't say what the reconciler does or how the struct is used.

rphillips replied:
Thank you @mikedanese. Added a stanza in commit 81b6c58

smarterclayton commented on contributors/design-proposals/apiserver-count-fix.md:
```diff
@@ -0,0 +1,114 @@
+
+| Kubernetes Release  | Quality | Description |
+| ------------- | ------------- | ----------- |
+| 1.9           | alpha         | <ul><li>Add a new reconciler</li><li>Add a command-line switch --new-reconciler</li><li>Add a command-line switch --old-reconciler</li></ul>
```
We generally don't add flags like this, it would usually be a type flag with an experimental prefix for alpha.

smarterclayton commented on contributors/design-proposals/apiserver-count-fix.md:
````diff
@@ -0,0 +1,114 @@
+}
+```
+
+Add a standard `kube-apiserver-endpoints` ConfigMap in the `default` namespace. The ConfigMap would be formed such that: 
````
The last proposed fix discussed adding an extension resource via crd instead of using config map.  Add a section discussing tradeoffs for that here

luxas replied:
Wasn't that for the full-blown APIServer status registration or do I recall incorrectly?
Anyway, I think that a ConfigMap is good here.
We've discussed layer violations of this kind before and it seems like that would be one.

smarterclayton replied:
I don't think it would be a layer violation, because this is the core api server (it's allowed to depend on apis).  We've already discussed it several times at an arch level.

lavalamp commented on contributors/design-proposals/apiserver-count-fix.md:
```diff
@@ -0,0 +1,114 @@
+
+Create a new `MasterEndpointReconciler` within master/controller.go.
+
+Add a `kube-apiserver-endpoints-config` ConfigMap in the `default` namespace. The duration found within the map would be configurable by admins without a recompile. The ConfigMap would include the following:
```
The namespace and configmap name should probably be flags :/

I would at least default to the `kube-system` namespace.

lavalamp replied:
Actually the name of this should probably be `kube-apiserver-config`; it should grow to include everything that is currently a flag. Endpoints don't need a special config thingy.

lavalamp replied:
I think someone may have been working on moving flags to a config map, you may want to coordinate with them. @mikedanese do you recall who was working on that?

lavalamp commented on contributors/design-proposals/apiserver-count-fix.md:
````diff
@@ -0,0 +1,114 @@
+
+```go
+ConfigMap{
+	"ip-2001-4860-4860--8888-443": "serialized JSON ControllerEndpointData",
````
Does anything other than a time need to be in the value? I'm not sure what benefit we get from cramming all the other stuff in there?

luxas replied:
Kind of agree; but at the same time we don't know what the future holds for us :)

lavalamp commented on contributors/design-proposals/apiserver-count-fix.md:
````diff
@@ -0,0 +1,114 @@
+Add a `kube-apiserver-endpoints-config` ConfigMap in the `default` namespace. The duration found within the map would be configurable by admins without a recompile. The ConfigMap would include the following:
+
+```go
+ConfigMap{
````
The name and namespace of the endpoint coordination config map should be stored in this configmap rather than be hard-coded.

lavalamp commented on contributors/design-proposals/apiserver-count-fix.md:
```diff
@@ -0,0 +1,114 @@
+The reconcile loop will expire endpoints that do not meet the duration. On
+each reconcile loop (the loop runs every 10 seconds currently):
+
+1. Retrieve `kube-apiserver-endpoints-config` ConfigMap (as configMap)
```
The config probably doesn't need to be gotten all that frequently, although it doesn't hurt anything, either.

luxas replied:
Is this GET on every reconcile loop or WATCH? Retrieve could mean both I guess

lavalamp commented on contributors/design-proposals/apiserver-count-fix.md:
```diff
@@ -0,0 +1,114 @@
+
+1. Retrieve `kube-apiserver-endpoints-config` ConfigMap (as configMap)
+1. Retrieve `kube-apiserver-endpoints` ConfigMap (as endpointMap)
+1. Update the `UpdateTimestamp` for the currently running API server
```
Can this be done via a PATCH?

lavalamp commented on contributors/design-proposals/apiserver-count-fix.md:
```diff
@@ -0,0 +1,114 @@
+1. Retrieve `kube-apiserver-endpoints-config` ConfigMap (as configMap)
+1. Retrieve `kube-apiserver-endpoints` ConfigMap (as endpointMap)
+1. Update the `UpdateTimestamp` for the currently running API server
+1. Remove all endpoints where the UpdateTimestamp is greater than `expire-duration` from the configMap.
```
I would do this in two requests rather than a single one, to avoid conflicts when multiple apiservers delete a record. Actually I'm not 100% sure it's a conflict if two requests delete the same thing, would a PATCH like that apply cleanly or give an error?

lavalamp commented on contributors/design-proposals/apiserver-count-fix.md:
````diff
@@ -0,0 +1,114 @@
+```
+
+The reconcile loop will expire endpoints that do not meet the duration. On
+each reconcile loop (the loop runs every 10 seconds currently):
````
Instead of a fixed 10s, this should probably be something like 50% to 80% of the expiration time, whatever that is.

lavalamp commented on contributors/design-proposals/apiserver-count-fix.md:
```diff
@@ -0,0 +1,114 @@
+| ------------- | ------------- | ----------- |
+| 1.9           | stable  | <ul><li>Change the logic in the current reconciler</li></ul>
+
+We could potentially reuse the old reconciler, but ignore the masterCount and change the logic to use the proposal from the previous section.
```
A first step might just be to count the number of entries in the config map and change the masterCount variable to that every time through the above loop (probably need to add a lock). That would probably be just a few line change.

rphillips commented:
Thank you all for the feedback! I'll add everything in.

fisherxu commented:
Have little doubts, why can‘t we add label on kube-apiserver's pod and through the selector in service sync the endpoint...

lavalamp commented:
> Have little doubts, why can‘t we add label on kube-apiserver's pod and through the selector in service sync the endpoint...

Yeah, this would definitely be the way to go if all installations of Kubernetes were self-hosted. Unfortunately that is not the case.

luxas commented on contributors/design-proposals/apiserver-count-fix.md:
````diff
@@ -0,0 +1,114 @@
+  namespace: default
+```
+
+### Refactor Old Reconciler
````
nit: Can you make it more obvious that these are the two current viable ways to do this? (and add one for the CRD thingy??)
like: `Possible solution 1: foo`, `Possible solution 2: bar`

luxas commented on contributors/design-proposals/apiserver-count-fix.md:
```diff
@@ -0,0 +1,114 @@
+
+## Prior Art
+
+[Security Labeller](https://github.com/coreos-inc/security-labeller/issues/18#issuecomment-320791878)
```
404 for me

luxas commented on contributors/design-proposals/apiserver-count-fix.md:
```diff
@@ -0,0 +1,114 @@
+
+1. [Overview](#overview)
+2. [Known Issues](#known-issues)
+3. [Proposal](#proposal)
```
Add sub-section for different strategies?

rphillips commented:
I've incorporated all the changes thus far.

There is still a pending question on what the payload is for  `KubeAPIServerEndpointConfigMap`, whether it should a string or a serialized structure. I have changed the field to be a serialization of time.Time, but I could swayed either way.

lavalamp commented:
I think some sort of config has to be in the nucleus. The nucleus needs to
be configured.

On Tue, Aug 22, 2017 at 12:56 PM, Clayton Coleman <notifications@github.com>
wrote:

> *@smarterclayton* commented on this pull request.
> ------------------------------
>
> In contributors/design-proposals/apiserver-count-fix.md
> <https://github.com/kubernetes/community/pull/939#discussion_r134585134>:
>
> > +| 1.10          | beta          | <ul><li>Turn on the new reconciler by default |</li></ul>
> +| 1.11          | stable        | <ul><li>Remove code for old reconciler</li><li>Remove --old-reconciler</li><li>Remove --new-reconciler</li><li>Remove --apiserver-count</li></ul>
> +
> +The MasterCountEndpointReconciler does not meet the current needs for durability of API Endpoint creation, deletion, or failure cases.
> +
> +Create a new `MasterEndpointReconciler` within master/controller.go.
> +
> +Add a `kube-apiserver-endpoints-config` ConfigMap in the `default` namespace. The duration found within the map would be configurable by admins without a recompile. The ConfigMap would include the following:
> +
> +```go
> +ConfigMap{
> +	"expire-duration": "1m", // golang duration (ns,us,ms,s,m,h)
> +}
> +```
> +
> +Add a standard `kube-apiserver-endpoints` ConfigMap in the `default` namespace. The ConfigMap would be formed such that:
>
> I don't think it would be a layer violation, because this is the core api
> server (it's allowed to depend on apis). We've already discussed it several
> times at an arch level.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/community/pull/939#discussion_r134585134>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAngloWSwKwho8w79p4-3U0cJZE9J4pBks5sazJ7gaJpZM4O625Y>
> .
>

smarterclayton commented on contributors/design-proposals/apiserver-count-fix.md:
```diff
@@ -0,0 +1,129 @@
+CRD's were considered for this proposal, but were not proposed due to
+constraints of having CRDs within core has not been clearly defined.
+Layering could also be an issue, so the proposal defined ConfigMaps as
+the current path forward.
```
@bgrant0607 when we discussed this before on whether we should be abusing configmaps for election we discussed a CRD as being acceptable for use within the api server.  

I think my preference at this point is not to abuse configmaps for high volume write traffic that is not related directly to the purpose of delivering configuration into a pod.  If we do not want to use the previously proposed mechanism of using storage with TTL (which I still think is the correct approach since it is how everyone else using etcd deals with leases) then we need to make a case that it is ok for a single component (the core kube-apiservers) to constantly write to a configmap, but that other components should not do so / are not affected by it.

bgrant0607 replied:
+1 to not abusing ConfigMap.

Extension mechanisms, such as CRD, will be part of the Nucleus.

lavalamp commented:
I almost suggested directly using etcd from apiserver, but I didn't because
I would like us to solve this problem in the way any user could. I don't
like having multiple mechanisms for solving the same problem.

Also etcd3's TTL is not as convenient as etcd2's was.

Do we really consider writing configmaps frequently as abuse? We recommend
using them for leader election, I thought. I don't see why we wouldn't
allow / want this use of a configmap.

​

smarterclayton commented:
I don't think we *want* users using configmap for election.  ConfigMap as a high write volume target that some components may ListAll/WatchAll gives me concern (garbage collector today, config map triggers in the future).  I'd rather have a resource that *no* one ever watches.  ConfigMap could easily be O(applications), which means any dense cluster SLO we have will potentially have to take into account O(applications) = 10k.  That makes configmap somewhat fraught.

The reason Brian and I discussed a CRD or new resources for this is because

1. CRD has to work
2. CRD is in core-isa
3. CRD can be arbitrarily sliced and ACLd

I don't want my apps able to write their own configmap (for security reasons).  I don't want to apply specific ACL magic rules to service accounts (all pods can update config maps that are linked in their pod), but we *can* do that for a CRD or core resource.

I agree with the sentiment of setting up a general mechanism for users to do this.  I do not agree with doing it on a general purpose like ConfigMap with more thought.  I prefer using something under the api layer (if this is api server specific) or something purpose built for frequent updates from pods with ACL implications taken into account (if this is not api server specific)

smarterclayton commented:
@kubernetes/sig-auth-feature-requests re: "pods that are able to write to config maps without ACL getting ugly" - we may be able to do this with RBAC, but if my pod can edit config maps I'd probably say by default that config map **should not** have other config in it.  At that point, it's not really a config map, it's "pod scratch coordination space".

lavalamp commented:
I agree re: configmaps for coordination should NOT also have other config
in them. This proposal has two config maps.

I will think about the other stuff you said and see if I can come up with
another alternative.

On Tue, Aug 22, 2017 at 1:18 PM, Clayton Coleman <notifications@github.com>
wrote:

> @kubernetes/sig-auth-feature-requests
> <https://github.com/orgs/kubernetes/teams/sig-auth-feature-requests> re:
> "pods that are able to write to config maps without ACL getting ugly" - we
> may be able to do this with RBAC, but if my pod can edit config maps I'd
> probably say by default that config map *should not* have other config in
> it. At that point, it's not really a config map, it's "pod scratch
> coordination space".
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/community/pull/939#issuecomment-324139903>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnglqexbwP5NTrkcUSWC5DCVJ48Qk98ks5sazesgaJpZM4O625Y>
> .
>

xiang90 commented:
@lavalamp 

> I almost suggested directly using etcd from apiserver, but I didn't because
I would like us to solve this problem in the way any user could. I don't
like having multiple mechanisms for solving the same problem.

user already could do this today for any applications running inside k8s as pods. k8s controller will update endpoints according to readiness.

we should not mix this proposal with leader election, since it is really not. it is a time-based liveness status for the "out-of-cluster" components to update endpoint. 


@smarterclayton 

>  I prefer using something under the api layer (if this is api server specific) or something purpose built for frequent updates from pods with ACL implications taken into account (if this is not api server specific)

Are you suggesting to enable this directly in the storage layer?

smarterclayton commented:
The storage layer can already be used for this -
https://github.com/openshift/origin/blob/master/pkg/cmd/server/election/lease_endpoint_reconciler.go

In talking with Xiang we also brought up another point - if we don't know
*exactly* the use case we want to solve for end users, we shouldn't half
create an API pattern here.  We should either design the pattern that
solves this generically, or specifically solve it under the api server.  I
lean towards us not having enough time to solve this generically, which is
why I still think that the kube-apiserver maintaining the minimal lease
based reconciler above is the right solution for the right now.

On Tue, Aug 22, 2017 at 4:36 PM, Xiang Li <notifications@github.com> wrote:

> @lavalamp <https://github.com/lavalamp>
>
> I almost suggested directly using etcd from apiserver, but I didn't because
> I would like us to solve this problem in the way any user could. I don't
> like having multiple mechanisms for solving the same problem.
>
> user already could do this today for any applications running inside k8s
> as pods. k8s controller will update endpoints according to readiness.
>
> we should not mix this proposal with leader election, since it is really
> not. it is a time-based liveness status for the "out-of-cluster" components
> to update endpoint.
>
> @smarterclayton <https://github.com/smarterclayton>
>
> I prefer using something under the api layer (if this is api server
> specific) or something purpose built for frequent updates from pods with
> ACL implications taken into account (if this is not api server specific)
>
> Are you suggesting to enable this directly in the storage layer?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/community/pull/939#issuecomment-324144418>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p0VO608XOJ6cj2iBzYl5_6HElfkOks5sazvcgaJpZM4O625Y>
> .
>

smarterclayton commented:
(also, the code linked above has been used in production for about a year on hundreds of clusters and we haven't had a problem yet, which makes me want to avoid solving a new problem)

xiang90 commented on contributors/design-proposals/apiserver-count-fix.md:
```diff
@@ -49,75 +50,27 @@ Endpoints would never become consistent.
+of API Endpoint creation, deletion, or failure cases.
 
-configmap.yml:
+Custom Resource Definitions and ConfigMaps were proposed, but since they are
```
I can understand ConfigMap has the watching issue. But CRD should not be watched globally. The problem with CRD , as far as I can tell,  is the not super clean layering, and potentially we might want something different that can handle both liveness and locking as a k8s native resource. And we are not sure yet.

rphillips commented:
@smarterclayton @lavalamp updated

I have the LeaseEndpointReconciler partially ported (clean build), but more work to do on it.

rphillips commented:
Is fixing kube-proxy is scope of this issue and eventual PR? or should it be a separate effort?

bgrant0607 commented on contributors/design-proposals/apiserver-count-fix.md:
```diff
@@ -0,0 +1,86 @@
+* The IP for endpoint ‘B’ is not
+removed from the Endpoints list
+
+There is logic within the
```
I assume aggregated apiservers won't do something similar?

liggitt replied:
aggregated servers can either delegate to service dns names (and therefore kube-proxy) or route to an IP selected from the endpoints for the service (and therefore take advantage of whatever mechanism removes unhealthy IP addresses from those endpoints)

lavalamp commented:
Aggregated API servers use the service concept to present themselves to the
system, so there's no need for them to do anything special here. (Yes,
perhaps technically it should just be endpoints, but the service concept
works and is easy.)

(We should make the main apiserver work the same way eventually, IMO)

On Wed, Aug 30, 2017 at 8:34 AM, Jordan Liggitt <notifications@github.com>
wrote:

> *@liggitt* commented on this pull request.
> ------------------------------
>
> In contributors/design-proposals/apiserver-count-fix.md
> <https://github.com/kubernetes/community/pull/939#discussion_r136105529>:
>
> > +instances within the Endpoints does not get cleaned up, or in the latter case
> +the endpoints start to flap.
> +
> +## Known Issues
> +
> +Each apiserver’s reconciler only cleans up for it's own IP. If a new
> +server is spun up at a new IP, then the old IP in the Endpoints list is
> +only reclaimed if the number of apiservers becomes greater-than or equal
> +to the masterCount. For example:
> +
> +* If the masterCount = 3, and there are 3 API servers running (named: A, B, and C)
> +* ‘B’ API server is terminated for any reason
> +* The IP for endpoint ‘B’ is not
> +removed from the Endpoints list
> +
> +There is logic within the
>
> aggregated servers can either delegate to service dns names (and therefore
> kube-proxy) or route to an IP selected from the endpoints for the service
> (and therefore take advantage of whatever mechanism removes unhealthy IP
> addresses from those endpoints)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/community/pull/939#discussion_r136105529>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnglmmiTgX3JudGoDB3ICiS3vAlsynoks5sdYEegaJpZM4O625Y>
> .
>

rphillips commented:
I have submitted a PR for this issue.

klausenbusk commented:
> Yeah, this would definitely be the way to go if all installations of Kubernetes were self-hosted. Unfortunately that is not the case.

Just a question. Is there anyway to disable the default reconciler? So I can try with `selector` already now.

lavalamp commented:
That's a great point, we should make the default behavior flag-disableable
so that it's possible to experiment with fully self hosted configurations.

On Thu, Aug 31, 2017 at 5:31 AM, Kristian Klausen <notifications@github.com>
wrote:

> Yeah, this would definitely be the way to go if all installations of
> Kubernetes were self-hosted. Unfortunately that is not the case.
>
> Just a question. Is there anyway to disable the default reconciler? So I
> can try with selector already now.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/community/pull/939#issuecomment-326281079>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnglmpyDSH0VhNGjHQTMp7XGkRCUfr8ks5sdqe8gaJpZM4O625Y>
> .
>

rphillips commented:
@lavalamp @klausenbusk  I added a 'none' reconciler to the PR to effectively be a noop.

lavalamp commented:
/approve
/lgtm

At the SIG meeting yesterday we collectively decided that using the storage interface directly was the least bad way of implementing this.

Note that other apiservers won't have this problem (they can use services & endpoints), it is unique to kube-apiserver.

luxas commented:
@lavalamp what "storage interface" are you regerring to in this context?

This will still use Endpoints, but in an other way, right?

lavalamp commented:
The storage.Interface go interface is what apiserver uses to talk to etcd.

On Thu, Aug 31, 2017 at 2:49 PM, Lucas Käldström <notifications@github.com>
wrote:

> @lavalamp <https://github.com/lavalamp> what "storage interface" are you
> regerring to in this context?
>
> This will still use Endpoints, but in an other way, right?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/community/pull/939#issuecomment-326429849>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAnglnqarsdOuD6IN8mXUUZJG_hVLUFyks5sdyp8gaJpZM4O625Y>
> .
>
